platform/lib/routers/search.js

```diff
@@ -0,0 +1,268 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const express = require('express');
+const fs = require('fs');
+const path = require('path');
+const config = require('@lib/config.js');
+const project = require('@lib/utils/project.js');
+const googleSearch = require('@lib/utils/googleSearch.js');
+
+// google custom search does not support a page size > 10
+const PAGE_SIZE = 10;
+// google custom search json api does not support loading more than 100 results
+const LAST_PAGE = 10;
+
+const COMPONENT_REFERENCE_DOC_PATTERN =
+    /^(?:https?:\/\/[^/]+)?(?:\/[^/]+)?\/documentation\/components\/(amp-[^/]+)/;
+
+const EXAMPLE_URL_PATH = '/documentation/examples/components/';
+const EXAMPLE_FOLDER_PATH = path.join(project.paths.DIST, '/examples/sources/components/');
+
+const REMOVE_HOST_PATTERN = /^https?:\/\/amp\.dev(?=\/)/;
+
+const PREVIEW_HOST = config.hosts.preview.base;
+const PLAYGROUND_URL = config.hosts.playground.base;
+
+// use the twitter title if available since it does not contain the site name
+const TITLE_META_TAG = 'twitter:title';
+const DESCRIPTION_META_TAG = 'twitter:description';
+
+// eslint-disable-next-line new-cap
+const examples = express.Router();
+
+const BUILD_IN_COMPONENTS = [
+  'amp-layout',
+  'amp-img',
+  'amp-pixel',
+];
+
+const IMPORTANT_INCLUDED_ELEMENTS = [
+  'amp-state',
+  'amp-list-load-more',
+  'amp-story-page',
+  'amp-story-grid-layer',
+  'amp-story-bookend',
+];
+
+const COMPONENT_VERSIONS_PATH = path.join(project.paths.GROW_POD,
+    '/extensions/amp-component-versions.json');
+
+function getAutosuggestComponents() {
+  let result = [];
+  const components = require(COMPONENT_VERSIONS_PATH);
+  for (const component in components) {
+    if (components.hasOwnProperty(component)) {
+      result.push(component);
+    }
+  }
+  result = result.concat(BUILD_IN_COMPONENTS);
+  result = result.concat(IMPORTANT_INCLUDED_ELEMENTS);
+  return result.sort();
+}
+
+const AUTOSUGGEST_COMPONENTS = getAutosuggestComponents();
+
+examples.get('/search/autosuggest', handleAutosuggestRequest);
+examples.get('/search/do', handleSearchRequest);
+
+function handleAutosuggestRequest(request, response) {
+  const items = ({
+    items: AUTOSUGGEST_COMPONENTS,
```

Comment:
nit: move items into the AUTOSUGGEST_COMPONENTS constant to avoid creating a new object every time.